domain/src/DomainAccessControlHandler.php

```diff
@@ -18,25 +18,61 @@ class DomainAccessControlHandler extends EntityAccessControlHandler {
    * {@inheritdoc}
    */
   public function checkAccess(EntityInterface $entity, $operation, AccountInterface $account = NULL) {
+    $account = $this->prepareUser($account);
     // Check the global permission.
     if ($account->hasPermission('administer domains')) {
       return AccessResult::allowed();
     }
-    if ($operation == 'view' && ($entity->status() || $account->hasPermission('access inactive domains'))) {
+    // @TODO: This may not be relevant.
+    if ($operation == 'create' && $account->hasPermission('create domains')) {
       return AccessResult::allowed();
     }
-    if ($operation == 'create' && $account->hasPermission('create domains')) {
+    // For view, we allow unless the domain is inactive.
+    $is_admin = $this->isDomainAdmin($entity, $account);
+    if ($operation == 'view' && ($entity->status() || $account->hasPermission('access inactive domains'))) {
       return AccessResult::allowed();
     }
-    // @TODO: assign users to domains and check.
-    if ($operation == 'update' && $account->hasPermission('edit assigned domains')) {
+    // For other operations, check that the user is a domain admin.
+    if ($operation == 'update' && $account->hasPermission('edit assigned domains') && $is_admin) {
       return AccessResult::allowed();
     }
-    // @TODO: assign users to domains and check.
-    if ($operation == 'delete' && $account->hasPermission('delete assigned domains')) {
+    if ($operation == 'delete' && $account->hasPermission('delete assigned domains') && $is_admin) {
       return AccessResult::allowed();
     }
     return AccessResult::forbidden();
   }
 
+  /**
+   * Checks if a user can administer a specific domain.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity to retrieve field data from.
+   * @param \Drupal\Core\Session\AccountInterface
+   *   The user account.
+   *
+   * @return boolean
+   */
+  public function isDomainAdmin(EntityInterface $entity, AccountInterface $account) {
+    $user = \Drupal::entityTypeManager()->getStorage('user')->load($account->id());
```

Comment:
I fumbled around with this for half an hour but can't get the create() method to fire at all, which causes fatal errors unless the EntityTypeManager defaults to NULL, which breaks the whole purpose. 

I must be missing something. Perhaps because there is a parent class that doesn't use this pattern, that method never fires? 

```
use Symfony\Component\DependencyInjection\ContainerInterface;

/**
 * Defines the access controller for the domain entity type.
 *
 * Note that this is not a node access check.
 */
class DomainAccessControlHandler extends EntityAccessControlHandler {

 /**
  * The Entity Type Manager
  *
  * @var \Drupal\Core\Entity\EntityTypeManagerInterface
  */
  protected $entityTypeManager;

  /**
   * Constructs an access control handler instance.
   *
   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
   *   The entity type definition.
   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
   */
  public function __construct(EntityTypeInterface $entity_type, EntityTypeManagerInterface $entity_type_manager = NULL) {
    parent::__construct($entity_type);
    $this->entityTypeManager = $entity_type_manager;
  }

  /**
   * {@inheritdoc}
   */
  public static function create(ContainerInterface $container, EntityTypeInterface $entity_type) {
    return new static(
      $entity_type,
      $container->get('entity_type.manager')
    );
  }
```